include/RAJA/policy/cuda/atomic.hpp

```diff
@@ -43,596 +51,549 @@ namespace RAJA
 {
-  return ::atomicMax((unsigned long long *)acc, value);
-}
-#endif
+  T old = cuda_atomicLoad(acc);
 
-#if __CUDA_ARCH__ >= 200
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicInc(T volatile *acc, T val)
-{
-  // See:
-  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicinc
-  return cuda_atomic_CAS_oper(acc, [=] __device__(T old) {
-    return ((old >= val) ? 0 : (old + 1));
-  });
-}
+  if (sc(old)) {
+    return old;
+  }
 
-// 32-bit unsigned atomicInc support by CUDA
-template <>
-RAJA_INLINE __device__ unsigned cuda_atomicInc<unsigned>(unsigned volatile *acc,
-                                                    unsigned value)
-{
-  return ::atomicInc((unsigned *)acc, value);
-}
+  T expected;
 
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicInc(T volatile *acc)
-{
-  return cuda_atomic_CAS_oper(acc,
-                                      [=] __device__(T a) { return a + 1; });
-}
+  do {
+    expected = old;
+    old = cuda_atomicCAS(acc, expected, oper(expected));
+  } while (!cuda_atomicCAS_equal(old, expected) && !sc(old));
 
-// 32-bit signed atomicAdd support by CUDA, used as backend for atomicInc
-template <>
-RAJA_INLINE __device__ int cuda_atomicInc<int>(int volatile *acc)
-{
-  return ::atomicAdd((int *)acc, (int)1);
+  return old;
 }
 
-// 32-bit unsigned atomicAdd support by CUDA, used as backend for atomicInc
-template <>
-RAJA_INLINE __device__ unsigned cuda_atomicInc<unsigned>(unsigned volatile *acc)
-{
-  return ::atomicAdd((unsigned *)acc, (unsigned)1);
-}
 
-// 64-bit unsigned atomicAdd support by CUDA, used as backend for atomicInc
-template <>
-RAJA_INLINE __device__ unsigned long long cuda_atomicInc<unsigned long long>(
-    unsigned long long volatile *acc)
+/*!
+ * Atomic addition
+ */
+using cuda_atomicAdd_builtin_types = list<
+      int
+     ,unsigned int
+     ,unsigned long long int
+     ,float
+#if __CUDA_ARCH__ >= 600
+     ,double
+#endif
+    >;
+
+template <typename T,
+          enable_if_is_none_of<T, cuda_atomicAdd_builtin_types>* = nullptr>
+RAJA_INLINE __device__ T cuda_atomicAdd(T *acc, T value)
 {
-  return ::atomicAdd((unsigned long long *)acc, (unsigned long long)1);
+  return cuda_atomicCAS(acc, [value] (T old) {
+    return old + value;
+  });
 }
 
-// 32-bit float atomicAdd support by CUDA, used as backend for atomicInc
-template <>
-RAJA_INLINE __device__ float cuda_atomicInc<float>(float volatile *acc)
+template <typename T,
+          enable_if_is_any_of<T, cuda_atomicAdd_builtin_types>* = nullptr>
+RAJA_INLINE __device__ T cuda_atomicAdd(T *acc, T value)
 {
-  return ::atomicAdd((float *)acc, (float)1);
+  return ::atomicAdd(acc, value);
 }
-#endif
 
-// 64-bit double atomicAdd support added for sm_60, used as backend for atomicInc
+
+/*!
+ * Atomic subtract
+ */
+using cuda_atomicSub_builtin_types = cuda_atomicAdd_builtin_types;
+
+using cuda_atomicSub_via_Sub_builtin_types = list<
+      int
```

Comment:
like here.... change to:

int,
unsigned int

??

Reply:
Fixed.